src/responses/Albany_SolutionTwoNormResponseFunction.cpp

```diff
@@ -158,13 +158,14 @@ evaluateDistParamHessVecProd_xx(
     const Teuchos::Array<ParamVec>& param_array,
     const Teuchos::RCP<Thyra_MultiVector>& Hv_dxdx)
 {
-  Teuchos::ScalarTraits<ST>::magnitudeType nrm = x->norm_2();
-  Teuchos::ScalarTraits<ST>::magnitudeType nrm3 = std::pow(nrm,3);
-
   if (!Hv_dxdx.is_null()) {
-    for (int j=0; j<Hv_dxdx->domain()->dim(); ++j) {
-      // Evaluate Hv_dxdx_j = v_j/||x|| - (v_j,x) x/||x||^3 where (.,.) is the usual dot product:
-      Thyra::V_StVpStV(Hv_dxdx->col(j).ptr(),1.0/nrm,*v->col(j),-1.0*x->dot(*v->col(j))/nrm3,*x);
+    Teuchos::ScalarTraits<ST>::magnitudeType nrm = x->norm_2();
+    if (nrm!=0.0) {
```

Comment:
I didn't mean to check if the nrm is indeed zero, only to put the norm related stuff inside the HV_dxdx nonnull test.

That said, it is true that if norm=0, then the formula looks bad. Should we throw if norm=0? My assumption is that all these derivatives, once multiplied with other queantities in Piro, ROL, or whatever, do not yield issues with norm=0 _in the analytical formulation_. Something along the line of a=x/|x|, b=x, then ab is well defined at 0, but a is not.

I don't think Albany can know how the Hessian derivatives will be used, so maybe we _should_ error out if the norm is 0?

Reply:
Not sure what's best honestly. It's true that if the norm is zero, which could happen, we don't necessarily want the code to throw an error, or even a NAN... A similar case happened when computing the derivative of sqrt(x), at x=0.

Reply:
Are you sure you don't want to throw if the derivatives are NaN or Inf? I'm ok with it, but then we need Piro to be fully prepared to handle "exceptional" cases... If Piro is not geared up to accept Inf or NaN (and decide whether to throw or patch up things), then we _should_ throw. Imho.

Reply:
OK, let's throw.